src/python/pants/backend/python/rules/pex.py

```diff
@@ -15,18 +15,65 @@
   DirectoriesToMerge,
   DirectoryWithPrefixToAdd,
 )
-from pants.engine.isolated_process import ExecuteProcessResult, MultiPlatformExecuteProcessRequest
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  MultiPlatformExecuteProcessRequest,
+)
+from pants.engine.legacy.structs import PythonTargetAdaptor, TargetAdaptor
 from pants.engine.platform import Platform, PlatformConstraint
-from pants.engine.rules import optionable_rule, rule
+from pants.engine.rules import RootRule, optionable_rule, rule
 from pants.engine.selectors import Get
+from pants.util.strutil import create_path_env_var
+
+
+@dataclass(frozen=True)
+class PexRequirements:
+  requirements: Tuple[str] = ()
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[TargetAdaptor], additional_requirements: Tuple[str]) -> 'PexRequirements':
+    all_target_requirements = []
+    for maybe_python_req_lib in adaptors:
+      # This is a python_requirement()-like target.
+      if hasattr(maybe_python_req_lib, 'requirement'):
+        all_target_requirements.append(str(maybe_python_req_lib.requirement))
+      # This is a python_requirement_library()-like target.
+      if hasattr(maybe_python_req_lib, 'requirements'):
+        for py_req in maybe_python_req_lib.requirements:
+          all_target_requirements.append(str(py_req.requirement))
+
+      all_target_requirements.extend(additional_requirements)
+    return PexRequirements(requirements=tuple(sorted(all_target_requirements)))
+
+@dataclass(frozen=True)
+class PexInterpreterContraints:
+  constraint_set: FrozenSet[str] = frozenset()
+
+  def generate_pex_arg_list(self) -> List[str]:
+    args = []
+    for constraint in self.constraint_set:
+      args.extend(["--interpreter-constraint", constraint])
+    return args
+
+  @classmethod
+  def create_from_adaptors(cls, adaptors: Tuple[PythonTargetAdaptor], python_setup: PythonSetup) -> 'PexInterpreterContraints':
+    interpreter_constraints = frozenset(
+      [constraint
+      for target_adaptor in adaptors
+      for constraint in python_setup.compatibility_or_constraints(
+        getattr(target_adaptor, 'compatibility', None)
+      )]
+    )
+    return PexInterpreterContraints(constraint_set=interpreter_constraints)
 
 
 @dataclass(frozen=True)
 class CreatePex:
   """Represents a generic request to create a PEX from its inputs."""
   output_filename: str
-  requirements: Tuple[str] = ()
-  interpreter_constraints: Tuple[str] = ()
+  requirements: PexRequirements = PexRequirements()
```

Comment:
What about instead making this `Optional[PexRequirements] = None`? That more clearly signals to me that there are no requirements by default, rather than having to rely on knowing that `PexRequirement()` will be empty by default.

Reply:
Hm, I have a slight preference for letting the type of this member be non-optional `PexRequirements`, since it means there's not two ways to represent "no requirements" (`None` and empty `PexRequirements()`) with identical semantics. Same for `PexInterpreterConstraints`.